File: 144-binary-tree-preorder-traversal/solution.cpp
#include <vector>
#include <deque>

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  std::vector<int> preorderTraversal(TreeNode *root)
  {
    std::vector<int> result;
    std::deque<const TreeNode *> work_deque{root};
    while (!work_deque.empty())
    {
      const auto *node = work_deque.back();
      work_deque.pop_back();
      if (node == nullptr)
      {
        continue;
      }

      result.emplace_back(node->val);
      work_deque.emplace_back(node->right);
      work_deque.emplace_back(node->left);
    }

    return result;
  }
};

File: 144-binary-tree-preorder-traversal/solution.py
from typing import Optional
from collections import deque


class TreeNode:
  def __init__(self, val=0, left: Optional['TreeNode'] = None, right: Optional['TreeNode'] = None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def preorderTraversal(self, root: Optional[TreeNode]) -> list[int]:
    result: list[int] = []
    work_deque: deque[TreeNode] = deque()
    curr = root
    while curr is not None or work_deque:
      while curr:
        result.append(curr.val)
        work_deque.append(curr)
        curr = curr.left

      node = work_deque.pop()
      curr = node.right

    return result
